modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/JavaParserUtils.java

```diff
@@ -0,0 +1,149 @@
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.loader.parser.java.utils;
+
+import org.mule.runtime.api.meta.MuleVersion;
+import org.mule.runtime.extension.api.exception.IllegalModelDefinitionException;
+import org.mule.runtime.module.extension.api.loader.java.type.ExtensionParameter;
+import org.mule.runtime.module.extension.api.loader.java.type.FieldElement;
+import org.mule.runtime.module.extension.api.loader.java.type.MethodElement;
+import org.mule.runtime.module.extension.api.loader.java.type.Type;
+import org.mule.runtime.module.extension.api.loader.java.type.TypeGeneric;
+import org.mule.runtime.module.extension.api.loader.java.type.WithAnnotations;
+import org.mule.sdk.api.annotation.MinMuleVersion;
+
+import org.slf4j.Logger;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+import static java.util.Optional.of;
+import static org.mule.sdk.api.util.SdkApiConstants.NON_ENFORCE_MIN_MULE_VERSION_CLASSES;
+import static org.slf4j.LoggerFactory.getLogger;
+
+public class JavaParserUtils {
+
+  private static final Logger LOGGER = getLogger(JavaParserUtils.class);
+  private static final String firstMuleVersion = "4.1.0";
+
+  private static final Set<String> primitiveTypeNames =
+      new HashSet<>(asList("byte", "short", "int", "long", "float", "double", "char", "boolean", "void"));
+  private static final List<Class<?>> denyList = NON_ENFORCE_MIN_MULE_VERSION_CLASSES;
+
+  public static String calculateFromClass(Type clazz) {
+    return calculateFromClass(clazz, new HashSet<>());
+  }
+
+  public static String calculateFromMethod(MethodElement<?> method) {
+    return calculateFromMethod(method, new HashSet<>());
+  }
+
+  public static String calculateFromParameter(ExtensionParameter parameter) {
+    return calculateFromField(parameter, new HashSet<>());
+  }
+
+  private static String calculateFromClass(Type clazz, Set<String> seenTypes) {
+    // TODO W-11938731 and W-11955483
+    if (clazz.getTypeName().startsWith("org.mule.extension.http") || clazz.getTypeName().startsWith("com.google.common")
+        || clazz.getTypeName().startsWith("java.")
+        || primitiveTypeNames.contains(clazz.getTypeName()) || seenTypes.contains(clazz.getTypeName())) {
+      return firstMuleVersion;
+    } else {
+      seenTypes.add(clazz.getTypeName());
+    }
+    // TODO: get MMV from pom
+    String calculatedMMV = firstMuleVersion;
+    // Look for the annotation at the class level
+    Optional<String> classLevelMMV = getMinMuleVersion(clazz);
+    if (classLevelMMV.isPresent()) { // We cut the algorithm short (only compare later against generics)
+      calculatedMMV = maxMMV(calculatedMMV, classLevelMMV.get());
+    } else { // We calculate from class inspection
```

Comment:
i don't think this is correct. If the user annotated the class with `@MinMuleVersion`, then for sure we should never resolve to a lower version. However, you still need to introspect fields, interfaces, etc. to make sure that the user is not forcing this to a LOWER version than it's really needed (in which case, this should be reported as a Problem)